### ISSUE TITLE
MobileNetV2 and YOLOv1

### ISSUE THREAD

**Issue by p30arena** [pjreddie/darknet#1832]
Hi!

I want to use MobileNetV2 instead of DarkNet,
But I can't get accuracy more than 0.003 after 10 hours of training.
I'm training the network on the WIDER FACE dataset.

total data is 11753
batch is 64
epoch is 210
optimizer is SGD with momentum 0.9 and decay 5e-4
learning rate is adaptive, based on the paper

I use only one bounding box
the grids are 15
the grid shape is 34
the image shape is 510

This is the model:
I apply Sigmoid activation on the output so I consider the predictions to be in normal form,
I've also normalized the ground truth beforehand.

```
i = tf.keras.layers.Input(shape=(img_shape, img_shape, 3))

x = tf.keras.applications.MobileNetV2(
  input_tensor=i,
  include_top=False,
  weights=None,
).output

x = tf.keras.layers.GlobalAveragePooling2D()(x)
o = tf.keras.layers.Dense((grid_shape * grid_shape * (nb_boxes * 5 + n_classes)),
                          activation='sigmoid', use_bias=True)(x)
o = tf.keras.layers.Reshape((grid_shape, grid_shape, (nb_boxes * 5 + n_classes)))(o)

model = tf.keras.Model(i, o)
```

This is the loss function (I don't apply SQRT on W and H, I mostly care about the big boxes):

```
def compute_iou(boxes1, boxes2):
  boxes1_t = tf.stack([boxes1[..., 0] - boxes1[..., 2] / 2.0,
                       boxes1[..., 1] - boxes1[..., 3] / 2.0,
                       boxes1[..., 0] + boxes1[..., 2] / 2.0,
                       boxes1[..., 1] + boxes1[..., 3] / 2.0],
                      axis=-1)

  boxes2_t = tf.stack([boxes2[..., 0] - boxes2[..., 2] / 2.0,
                       boxes2[..., 1] - boxes2[..., 3] / 2.0,
                       boxes2[..., 0] + boxes2[..., 2] / 2.0,
                       boxes2[..., 1] + boxes2[..., 3] / 2.0],
                      axis=-1)

  lu = tf.maximum(boxes1_t[..., :2], boxes2_t[..., :2])
  rd = tf.minimum(boxes1_t[..., 2:], boxes2_t[..., 2:])

  intersection = tf.maximum(0.0, rd - lu)
  inter_square = intersection[..., 0] * intersection[..., 1]

  square1 = boxes1[..., 2] * boxes1[..., 3]
  square2 = boxes2[..., 2] * boxes2[..., 3]

  union_square = tf.maximum(square1 + square2 - inter_square, 1e-10)
  return tf.clip_by_value(inter_square / union_square, 0.0, 1.0)


#
# The loss function orient the backpropagation algorithm toward the best direction.
# It does so by outputting a number. The larger the number, the further we are from a correct solution.
# Keras also accept that we output a tensor. In that case it will just sum all the numbers to get a single number.
#
@tf.function
def Yolo_Loss(y_true, y_pred):
  K = tf.keras.backend

  y_pred_conf = y_pred[..., :nb_boxes]
  y_true_conf = y_true[..., :nb_boxes]
  y_pred_class = y_pred[..., 5 * nb_boxes:]
  y_true_class = y_true[..., 5 * nb_boxes:]

  # sigmoid output is between 0 and 1, so we consider it to be in the normal form
  pred_boxes_norm = K.reshape(y_pred[..., nb_boxes:5 * nb_boxes], (-1, grid_shape, grid_shape, nb_boxes, 4))
  true_boxes_norm = K.reshape(y_true[..., nb_boxes:5 * nb_boxes], (-1, grid_shape, grid_shape, nb_boxes, 4))

  iou = compute_iou(true_boxes_norm, pred_boxes_norm)

  predictor_mask = tf.reduce_max(iou, axis=3, keepdims=True)
  predictor_mask = tf.cast(iou >= predictor_mask,
                           tf.float32) * y_true_conf
  noobj_mask = tf.ones_like(predictor_mask) - predictor_mask

  # Computing the class loss
  class_loss = tf.reduce_mean(tf.reduce_sum(tf.square(
    y_true_conf * (y_true_class - y_pred_class)), axis=[1, 2, 3]))

  # Computing the confidence loss
  obj_loss = tf.reduce_mean(tf.reduce_sum(
    tf.square(predictor_mask * (y_pred_conf - iou)), axis=[1, 2, 3]))
  noobj_loss = tf.reduce_mean(tf.reduce_sum(
    tf.square(noobj_mask * y_pred_conf), axis=[1, 2, 3]))

  # Computing the localization loss
  predictor_mask = predictor_mask[..., None]
  loc_loss = tf.reduce_mean(tf.reduce_sum(tf.square(
    predictor_mask * (true_boxes_norm - pred_boxes_norm)), axis=[1, 2, 3]))

  return 5 * loc_loss + 1.0 * obj_loss + 0.5 * noobj_loss + 1.0 * class_loss

```

**Comment by p30arena** [comment]
If I replace the dense layer with Convolutional Layers, I reach more accuracy about 30% in 10 Hours.

Is there a problem with my loss function?

```
i = tf.keras.layers.Input(shape=(img_shape, img_shape, 3))

x = tf.keras.applications.MobileNetV2(
  input_tensor=i,
  include_top=False,
  weights=None,
).layers[-4].output

 x = tf.keras.layers.MaxPooling2D(strides=1)(x)

face_prob = block(x, n_filters=1280)
xy_2d = block(x, n_filters=1280)
wh_2d = block(x, n_filters=1280)

face_prob = tf.keras.layers.Conv2D(2, kernel_size=1, use_bias=True)(face_prob)
prob = tf.keras.layers.Activation(activation='sigmoid')(face_prob[..., 0])
face_class = tf.keras.layers.Activation(activation='sigmoid')(face_prob[..., 1])

xy_2d = tf.keras.layers.Conv2D(2, kernel_size=1, use_bias=True)(xy_2d)
x_center = tf.keras.layers.Activation(activation='sigmoid')(xy_2d[..., 0])
y_center = tf.keras.layers.Activation(activation='sigmoid')(xy_2d[..., 1])

wh_2d = tf.keras.layers.Conv2D(2, kernel_size=1, use_bias=True)(wh_2d)
w = tf.keras.layers.Activation(activation='sigmoid')(wh_2d[..., 0])
h = tf.keras.layers.Activation(activation='sigmoid')(wh_2d[..., 1])

o = tf.keras.layers.concatenate([prob, x_center, y_center, w, h, face_class])
o = tf.keras.layers.Reshape((grid_shape, grid_shape, (nb_boxes * 5 + n_classes)))(o)

model = tf.keras.Model(i, o)
```

**Comment by p30arena** [comment]
I borrowed some concepts from YOLOv2

I changed the image size form 510 to 416 and moved the activations to the loss function
grids are 13
batch is 32

around epoch 23 the accuracy is now 22%

```
i = tf.keras.layers.Input(shape=(img_shape, img_shape, 3))

x = tf.keras.applications.MobileNetV2(
  input_tensor=i,
  include_top=False,
  weights=None,
).output

o = tf.keras.layers.Conv2D((nb_boxes * 5 + n_classes),
                           kernel_size=1, padding='same', activation='linear')(x)
o = tf.keras.layers.Reshape((grid_shape, grid_shape, (nb_boxes * 5 + n_classes)))(o)

model = tf.keras.Model(i, o)
```

```
K = tf.keras.backend
  y_pred = tf.identity(y_pred, name='y_pred')

  y_pred_conf = tf.nn.sigmoid(y_pred[..., :nb_boxes])
  y_true_conf = y_true[..., :nb_boxes]
  y_pred_class = tf.nn.sigmoid(y_pred[..., 5 * nb_boxes:])
  y_true_class = y_true[..., 5 * nb_boxes:]

  # sigmoid output is between 0 and 1, so we consider it to be in the normal form
  pred_boxes_norm = K.reshape(tf.nn.sigmoid(y_pred[..., nb_boxes:5 * nb_boxes]),
                              (-1, grid_shape, grid_shape, nb_boxes, 4))
  true_boxes_norm = K.reshape(y_true[..., nb_boxes:5 * nb_boxes], (-1, grid_shape, grid_shape, nb_boxes, 4))

...
```
